server/src/main/java/io/unitycatalog/server/service/credential/aws/AwsPolicyGenerator.java

```diff
@@ -0,0 +1,122 @@
+package io.unitycatalog.server.service.credential.aws;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import io.unitycatalog.server.service.credential.CredentialContext;
+import lombok.SneakyThrows;
+import org.apache.iceberg.exceptions.NotAuthorizedException;
+
+import java.net.URI;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class AwsPolicyGenerator {
+
+  static final List<String> SELECT_ACTIONS = List.of("s3:GetO*");
+  static final List<String> UPDATE_ACTIONS =
+    List.of("s3:GetO*", "s3:PutO*", "s3:DeleteO*", "s3:*Multipart*");
+
+  static final String POLICY_STATEMENT =
+    """
+    Version: 2012-10-17
+    Statement: []
+    """;
+
+  static final String BUCKET_STATEMENT =
+    """
+    Effect: Allow
+    Action:
+      - s3:ListBucket
+    Resource: []
+    Condition:
+      StringLike:
+        "s3:prefix": []
+    """;
+
+  static final String OPERATION_STATEMENT =
+    """
+    Effect: Allow
+    Action: []
+    Resource: []
+    """;
+
+  private static final ObjectMapper JSON_MAPPER = new ObjectMapper();
+  private static final ObjectMapper YAML_MAPPER = new ObjectMapper(new YAMLFactory());
+
+  // This can support generating a policy across multiple buckets and paths, however, the assumed role
+  // the policy is applied to for a scoped-session needs to have access across those buckets
+  @SneakyThrows
+  public static String generatePolicy(Set<CredentialContext.Privilege> privileges, List<String> locations) {
```

Comment:
brought to my attention that there is now an IAM policybuilder library that may make some of this easier and cleaner, maybe can look at using that in an follow-up PR (or update this PR if really necessary, though what is here has already been testing)